File: python/src/avet.py
# AVET: attribute → value → set of entity_ids
# Fast for: "which entities have attribute A with value V?"

def index_add(index, datom):
    attr = index.get(datom.attr_name, {})
    existing = attr.get(datom.value, set())
    return {**index, datom.attr_name: {**attr, datom.value: existing | {datom.entity_id}}}


def index_remove(index, datom):
    attr = index.get(datom.attr_name, {})
    new_set = attr.get(datom.value, set()) - {datom.entity_id}
    return {**index, datom.attr_name: {**attr, datom.value: new_set}}


def index_get(index, attr_name, value=None):
    attr = dict(index.get(attr_name, {}))
    if value is None:
        return attr
    return attr.get(value, set())

File: clojure/src/circle_db/avet.clj
(ns circle-db.avet)
;; AVET: attribute → value → set of entity-ids
;; Fast for: "which entities have attribute A with value V?"

(defn index-add [index datom]
  (update-in index [(:attr-name datom) (:value datom)] (fnil conj #{}) (:entity-id datom)))

(defn index-remove [index datom]
  (update-in index [(:attr-name datom) (:value datom)] disj (:entity-id datom)))

(defn index-get
  ([index attr-name]
   (get index attr-name {}))
  ([index attr-name value]
   (get-in index [attr-name value] #{})))
